src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java

```diff
@@ -112,20 +136,88 @@ private Empty() {
 
         private final Set<String> names = new HashSet<>();
         private final List<AggregatorFactory> factories = new ArrayList<>();
+        private final List<ReducerFactory> reducerFactories = new ArrayList<>();
 
-        public Builder add(AggregatorFactory factory) {
+        public Builder addAggregator(AggregatorFactory factory) {
             if (!names.add(factory.name)) {
-                throw new ElasticsearchIllegalArgumentException("Two sibling aggregations cannot have the same name: [" + factory.name + "]");
+                throw new ElasticsearchIllegalArgumentException("Two sibling aggregations cannot have the same name: [" + factory.name
+                        + "]");
             }
             factories.add(factory);
             return this;
         }
 
+        public Builder addReducer(ReducerFactory reducerFactory) {
+            this.reducerFactories.add(reducerFactory);
+            return this;
+        }
+
         public AggregatorFactories build() {
-            if (factories.isEmpty()) {
+            if (factories.isEmpty() && reducerFactories.isEmpty()) {
                 return EMPTY;
             }
-            return new AggregatorFactories(factories.toArray(new AggregatorFactory[factories.size()]));
+            List<ReducerFactory> orderedReducers = resolveReducerOrder(this.reducerFactories, this.factories);
+            return new AggregatorFactories(factories.toArray(new AggregatorFactory[factories.size()]), orderedReducers);
+        }
+
+        /*
+         * L ← Empty list that will contain the sorted nodes while there are
+         * unmarked nodes do select an unmarked node n visit(n) function
+         * visit(node n) if n has a temporary mark then stop (not a DAG) if n is
+         * not marked (i.e. has not been visited yet) then mark n temporarily
+         * for each node m with an edge from n to m do visit(m) mark n
+         * permanently unmark n temporarily add n to head of L
```

Comment:
This was supposed to be removed, was only useful while I implemented the ordering logic